cmd/thanos/query-frontend.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package main
+
+import (
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/querier/frontend"
+	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+	"github.com/go-kit/kit/log"
+	"github.com/oklog/run"
+	"github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/route"
+	"gopkg.in/alecthomas/kingpin.v2"
+
+	v1 "github.com/thanos-io/thanos/pkg/api/queryfrontend"
+	"github.com/thanos-io/thanos/pkg/component"
+	"github.com/thanos-io/thanos/pkg/extprom"
+	"github.com/thanos-io/thanos/pkg/prober"
+	"github.com/thanos-io/thanos/pkg/queryfrontend"
+	httpserver "github.com/thanos-io/thanos/pkg/server/http"
+)
+
+type queryFrontendConfig struct {
+	http             httpConfig
+	queryRangeConfig queryRangeConfig
+
+	// TODO(yeya24): use this after promql engine is added
+	defaultEvaluationInterval time.Duration
+	downstreamURL             string
+	compressResponses         bool
+}
+
+type queryRangeConfig struct {
+	respCacheConfig     responseCacheConfig
+	cacheResults        bool
+	splitInterval       time.Duration
+	disableStepAlign    bool
+	maxRetries          int
+	maxQueryParallelism int
+	maxQueryLength      time.Duration
+}
+
+type responseCacheConfig struct {
+	cacheMaxFreshness time.Duration
+}
+
+func (c *responseCacheConfig) registerFlag(cmd *kingpin.CmdClause) {
+	cmd.Flag("query-range.response-cache-max-freshness", "Most recent allowed cacheable result per-tenant, to prevent caching very recent results that might still be in flux.").
+		Default("1m").DurationVar(&c.cacheMaxFreshness)
+}
+
+func (c *queryRangeConfig) registerFlag(cmd *kingpin.CmdClause) {
+	c.respCacheConfig.registerFlag(cmd)
+
+	cmd.Flag("query-range.cache-results", "Cache query range results.").Default("false").
+		BoolVar(&c.cacheResults)
+
+	cmd.Flag("query-range.split-interval", "Split queries by an interval and execute in parallel, 0 disables it.").
+		Default("24h").DurationVar(&c.splitInterval)
+
+	cmd.Flag("query-range.max-retries-per-request", "Maximum number of retries for a single request; beyond this, the downstream error is returned.").
+		Default("5").IntVar(&c.maxRetries)
+
+	cmd.Flag("query-range.disable-step-align", "Disable aligning incoming queries' start and end with their step.").
```

Comment:
Let's remove this option and only expose this if someone will need. In practice you really want true. Plus Grafana does that by default